distage/distage-framework-docker/src/main/scala/izumi/distage/docker/healthcheck/ContainerHealthCheck.scala

```diff
@@ -28,41 +28,57 @@ trait ContainerHealthCheck[Tag] {
 }
 
 object ContainerHealthCheck {
-  def portCheck[T]: ContainerHealthCheck[T] = new TCPContainerHealthCheck[T]
-  def httpGetCheck[T](port: DockerPort): ContainerHealthCheck[T] = portCheck.combineOnPorts(new HttpGetCheck(_, port, useHttps = false))
-  def httpsGetCheck[T](port: DockerPort): ContainerHealthCheck[T] = portCheck.combineOnPorts(new HttpGetCheck(_, port, useHttps = true))
-  def postgreSqlProtocolCheck[T](port: DockerPort, user: String, password: String): ContainerHealthCheck[T] =
+  def portCheck[T]: ContainerHealthCheck[T] = {
+    new TCPContainerHealthCheck[T]
+  }
+
+  def httpGetCheck[T](port: DockerPort): ContainerHealthCheck[T] = {
+    portCheck.combineOnPorts(new HttpGetCheck(_, port, useHttps = false))
+  }
+  def httpsGetCheck[T](port: DockerPort): ContainerHealthCheck[T] = {
+    portCheck.combineOnPorts(new HttpGetCheck(_, port, useHttps = true))
+  }
+
+  def postgreSqlProtocolCheck[T](port: DockerPort, user: String, password: String): ContainerHealthCheck[T] = {
     portCheck.combineOnPorts(new PostgreSqlProtocolCheck(_, port, user, password))
+  }
 
   def succeed[T]: ContainerHealthCheck[T] = (_, _) => HealthCheckResult.Available
 
   sealed trait HealthCheckResult
   object HealthCheckResult {
-    case object Available extends HealthCheckResult
-    case object Unavailable extends HealthCheckResult
+    sealed trait BadHealthcheck extends HealthCheckResult
+    case object Unavailable extends BadHealthcheck
+
+    sealed trait GoodHealthcheck extends HealthCheckResult
+    case object Available extends GoodHealthcheck
+
     final case class AvailableOnPorts(
-      availablePorts: VerifiedContainerConnectivity,
+      availablePorts: AvailablePorts,
       unavailablePorts: UnavailablePorts,
       unverifiedPorts: Set[DockerPort],
-      requiredPortsAccessible: Boolean,
-    ) extends HealthCheckResult
-  }
+      allTCPPortsAccessible: Boolean,
+    ) extends GoodHealthcheck
 
-  final case class UnavailablePorts(unavailablePorts: Map[DockerPort, List[(AvailablePort, Option[Throwable])]]) extends AnyVal {
-    override def toString: String = s"unavailable=$unavailablePorts"
+    final case class UnavailableWithMeta(
+      unavailablePorts: UnavailablePorts,
+      unverifiedPorts: Set[DockerPort],
+    ) extends BadHealthcheck
   }
 
-  final case class VerifiedContainerConnectivity(availablePorts: Map[DockerPort, NonEmptyList[AvailablePort]]) extends AnyVal {
+  final case class AvailablePorts(availablePorts: NonEmptyMap[DockerPort, NonEmptyList[AvailablePort]]) {
```

Comment:
Being non-empty doesn't actually make this any safer since the `port` argument may still be missing from the map in all of the functions below.